File: 2015/qualification/omino/Omino.hs
module Omino where

import Data.List.Split (splitOn)

data Omino = Omino Int Int Int

main = do
        t <- getLine
        contents <- getContents
        let problems = (parse $ lines contents)
        mapM_ (putStrLn . presentCase) (zip [1..] . solveAll $ problems)

parse :: [String] -> [Omino]
parse [] = []
parse (line:lines) = let [x, r, c] = splitOn " " line
                     in (Omino (read x) (read r) (read c)) : parse lines

presentCase :: (Int, Bool) -> String
presentCase (i, result)  = "Case #" ++ show i ++ ": " ++ presentResult result

presentResult :: Bool -> String
presentResult True  = "GABRIEL"
presentResult False = "RICHARD"

solveAll :: [Omino] -> [Bool]
solveAll = map solve

solve :: Omino -> Bool
solve o@(Omino x r c)
        | ((area o) `mod` x) /= 0        = False
        | area o >= 7                    = False
        | area o == 6 && (min r c) <=  3 = False
        | area o == 5 && ((min r c) <= 2 || ((min r c) == 3 && (max r c) == 5)) = False
        | area o == 4 && (min r c) <= 2  = False
        | area o == 3 && (min r c) <= 1  = False
        | otherwise                      = True

area :: Omino -> Int
area (Omino _ r c) = r * c

File: 2015/qualification/omino/omino.rb
def solve(x, r, c)
  area = r * c

  case x
  when 1
    "GABRIEL"
  when 2
    if area.even?
      "GABRIEL"
    else
      "RICHARD"
    end
  when 3
    if area != 3 && area % x == 0
      "GABRIEL"
    else
      "RICHARD"
    end
  when 4
    if area == 12 || area == 16
      "GABRIEL"
    else
      "RICHARD"
    end
  end
end

T = STDIN.gets.to_i

T.times.collect { |i|
  x, r, c = STDIN.gets.split(' ').map(&:to_i)

  puts "Case ##{ i + 1 }: #{ solve(x, r, c) }"

}
